### PR TITLE
Tweak Private Sbbf::new_with_ndv_fpp Documentation

### LINKED ISSUE #5708: Inconsistency between code and comment
I found some inconsistency between code and comment.
In arrow-rs-master/parquet/src/bloom_filter/mod.rs, the comment “panic if `fpp` is greater than 1.0 or less than 0.0” indicates that 0.0 <= fpp <= 1.0 while in the code "!(0.0..1.0).contains(&fpp)" indicates that 0.0 <= fpp <1.0.

```rust
/// Will panic if `fpp` is greater than 1.0 or less than 0.0.
    pub(crate) fn new_with_ndv_fpp(ndv: u64, fpp: f64) -> Result<Self, ParquetError> {
        if !(0.0..1.0).contains(&fpp) {
            return Err(ParquetError::General(format!(
                "False positive probability must be between 0.0 and 1.0, got {fpp}"
            )));
```
and some out-of-bounds constraints mentioned in comment should be checked in code,  I don't think“out-of-bound” crash is the behavior we expect. So I think use assert! to check whether the index is out-of-bound before using it is a better way.
In arrow-rs-master/arrow-array/src/array/byte_array.rs, I think should add assert!(i + 1 < offsets.len()) before using i directly as index.
```rust
   /// # Panics
    /// Panics if index `i` is out of bounds.
    #[inline]
    pub fn value_length(&self, i: usize) -> T::Offset {
        let offsets = self.value_offsets();
        offsets[i + 1] - offsets[i]
    }
```
In arrow-rs-master/arrow-array/src/array/byte_array.rs
```rust
    /// # Panic
    /// This function panics if there is an invalid index,
    /// i.e. `index` >= the number of source arrays
    /// or `end` > the length of the `index`th array
    pub fn extend(&mut self, index: usize, start: usize, end: usize) {
        let len = end - start;
        (self.extend_null_bits[index])(&mut self.data, start, len);
        (self.extend_values[index])(&mut self.data, index, start, len);
        self.data.len += len;
    }
```
In arrow-rs-master/arrow-schema/src/schema.rs
```rust
    /// # Panics
    /// Panics if index out of bounds
    pub fn field(&mut self, idx: usize) -> &FieldRef {
        &mut self.fields[idx]
    }
```

### PR DESCRIPTION
**Pull request by YichiZhang0613**
Closes #5708.

I solve some inconsistency between code and comment and add some checks for boundary constraints mentioned in comment.

### CODE REVIEW

File: arrow-array/src/array/byte_array.rs [hunk 1]
```diff
@@ -232,6 +232,7 @@ impl<T: ByteArrayType> GenericByteArray<T> {
     #[inline]
     pub fn value_length(&self, i: usize) -> T::Offset {
         let offsets = self.value_offsets();
+        assert!(i + 1 < offsets.len());
```
Comment (alamb) [inline arrow-array/src/array/byte_array.rs:235]: I think this check is unecessary as `offsets[i + 1]` will panic if the access is out of bounds

**Review by alamb — COMMENTED** [review]
Thanks for this contribution @YichiZhang0613  🙏 

I think the asserts in this PR are unecessary as Rust automatically checks bounds on array accesses (and panic's if it is out of bounds)

File: arrow-data/src/transform/mod.rs [hunk 2]
```diff
@@ -645,6 +645,8 @@ impl<'a> MutableArrayData<'a> {
     /// or `end` > the length of the `index`th array
     pub fn extend(&mut self, index: usize, start: usize, end: usize) {
         let len = end - start;
+        assert!(index < self.extend_null_bits.len());
```
Comment (tustvold) [inline arrow-data/src/transform/mod.rs:648]: These are also unnecessary

File: parquet/src/bloom_filter/mod.rs [hunk 3]
```diff
@@ -179,7 +179,7 @@ fn num_of_bits_from_ndv_fpp(ndv: u64, fpp: f64) -> usize {
 
 impl Sbbf {
     /// Create a new [Sbbf] with given number of distinct values and false positive probability.
-    /// Will panic if `fpp` is greater than 1.0 or less than 0.0.
+    /// Will panic if `fpp` is greater than or equal to 1.0 or less than 0.0.
```
Comment (tustvold) [inline parquet/src/bloom_filter/mod.rs:182]:
```suggestion
    /// Will return an error if `fpp` is greater than or equal to 1.0 or less than 0.0.
```

**Pull request merged**